lib/api-helper/listener/executeCancel.ts

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright © 2018 Atomist, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+    addressEvent,
+    AutomationContextAware,
+    QueryNoCacheOptions,
+} from "@atomist/automation-client";
+import { codeLine } from "@atomist/slack-messages";
+import * as _ from "lodash";
+import { CancelOptions } from "../../api/goal/common/Cancel";
+import { ExecuteGoal } from "../../api/goal/GoalInvocation";
+import { isGoals } from "../../api/goal/Goals";
+import { GoalRootType } from "../../api/goal/SdmGoalMessage";
+import {
+    SdmGoalByShaAndBranch,
+    SdmGoalState,
+} from "../../typings/types";
+import { sumSdmGoalEvents } from "../goal/fetchGoalsOnCommit";
+
+/**
+ * Cancel any pending goals that are on the previous commit of the goal's branch
+ * @param options
+ * @param name
+ */
+export function executeCancelGoalSets(options: CancelOptions, name: string): ExecuteGoal {
+    return async gi => {
+        const registration = gi.configuration.name;
+
+        const goals = await gi.context.graphClient.query<SdmGoalByShaAndBranch.Query, SdmGoalByShaAndBranch.Variables>({
+            name: "SdmGoalByShaAndBranch",
+            variables: {
+                sha: gi.sdmGoal.push.before.sha,
+                branch: gi.sdmGoal.branch,
+                repo: gi.sdmGoal.repo.name,
+                owner: gi.sdmGoal.repo.owner,
+                providerId: gi.sdmGoal.repo.providerId,
+                uniqueNames: _.uniq(_.flatten(options.goals.map(g => {
+                    if (isGoals(g)) {
+                        return g.goals.map(gg => gg.uniqueName);
+                    } else {
+                        return g.uniqueName;
+                    }
+                }))),
+            },
+            options: QueryNoCacheOptions,
+        });
+
+        if (goals && goals.SdmGoal && goals.SdmGoal.length > 0) {
+            const currentGoals = sumSdmGoalEvents(goals.SdmGoal as any) as SdmGoalByShaAndBranch.SdmGoal[];
+            const cancelableGoals = currentGoals.filter(g => provenanceFilter(g, registration))
+                .filter(g => options.goalFilter(g as any));
```

Comment:
Is it possible to create a Cancel-like goal without using `new Cancel()`? If so, `goalFilter` might be `undefined`.

Reply:
Good point, fixing.